---
title: "RNA-seq (DESEQ2)"
output: html_document
date: "2025-02-01"
---

## 0- My session info

```{r}
sessionInfo()
```

## 1- Loading libraries

```{r Libraries}
library(readr)
library(org.Hs.eg.db)
library(vsn)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(ggfortify)
library(ggrepel)
library(ComplexHeatmap)
library(circlize)
library(rgl)

#setting directory
setwd("E:/1.Fresh Grad/02_EgComBio2023/MODA RNA_seq")
```

## 2- Loading the data and metadata

```{r Data}
read_counts <- read.delim("E:/1.Fresh Grad/02_EgComBio2023/MODA RNA_seq/GSE275290_raw_counts.tsv")
metadata <- read.csv("E:/1.Fresh Grad/02_EgComBio2023/MODA RNA_seq/Phenotable.csv")

```

## 3- Exploratory Data Analysis: Box plot and Histogram

```{r Exploration}
boxplot(log2(read_counts[,-1]+1), main="Exploratory Box plot", ylab="log2_count", las=2)
hist(as.matrix(log2(read_counts[,-1]+1)), main="Exploratory Histogram", xlab="Sample", ylab="log2_count", breaks=50)
```

## 4- Gene Annotation (differs according to dataset publisher uploading format)

```{r Annotation}
# Mapping gene IDs to gene symbols
gene_ids <- as.character(read_counts$GeneID)  # Ensure gene IDs are characters
# Convert ENTREZ IDs to gene symbols using org.Hs.eg.db
gene_symbol <- mapIds(
  org.Hs.eg.db, keys = gene_ids, keytype = "ENTREZID",
  column = "SYMBOL", multiVals = "first")
# Create a data frame with gene IDs and corresponding gene symbols
gene_df <- data.frame(
  GeneID = names(gene_symbol),
  Gene_symbol = as.vector(gene_symbol),
  stringsAsFactors = FALSE)
# Merge the gene symbols with the raw counts data
data <- merge(read_counts, gene_df, by="GeneID", all.x=TRUE)
data <- data %>%
  dplyr::select(Gene_symbol, everything(), -GeneID)
```

## 5- Preprocessing
### 5- (a) Preprocess metadata

```{r Preprocesing I}
meta <- metadata %>%
  select(Sample.Name, treatment) %>%
  rename(sampleid = Sample.Name, Condition = treatment)

meta <- meta %>%
  mutate(Condition = case_when(
    Condition == "FBZ for 48 hours" ~ "FBZ",
    Condition == "DMSO; Vehicle" ~ "Control",
    TRUE ~ Condition))
```

### 5- (b) Preprocess expression matrix

```{r Preprocesing II}
# Remove rows with NA values
sum(is.na(data))
data <- na.omit(data)

# Check for duplicated gene symbols
sum(duplicated(data$Gene_symbol))

#If yes!! Aggregate expression data by gene symbol
exp_data <- data %>% 
  dplyr::select(-Gene_symbol)
is.numeric(exp_data)  # Check if the data is numeric
exp_data_agg <- aggregate(exp_data, by=list(data$Gene_symbol), FUN=mean)

sum(duplicated(exp_data_agg$Group.1))  # Check for duplicated gene symbols after aggregation

# Set row names and remove the Group.1 column
row.names(exp_data_agg) <- exp_data_agg$Group.1
exp_data_agg <- exp_data_agg[,-1]

# Remove genes with zero variance
varrow <- apply(exp_data_agg, 1, var, na.rm=TRUE)
cons_var <- (varrow == 0 | is.na(varrow))
exp_data_agg <- exp_data_agg[!cons_var,]
```

## 6- Follow the exploratory analysis on the processed data

### Perform PCA on the filtered data

```{r PCA}
pca <- prcomp(t(log2(exp_data_agg + 1)), scale. = TRUE)
plot(pca$x[,1], pca$x[,2], main = "PCA of Filtered Raw Counts", xlab = "PC1", ylab = "PC2")
autoplot(pca, data = meta, colour = 'Condition',frame = T,label = T, label.size = 3,shape="Condition")

```

## 7- Differential expression analysis

## ----------DESeq2 package-----------

#### [**Note:**]{.underline} DESeq2 uses raw counts for its internal statistical model so no normalization is required

```{r DEseq2}
exp <- exp_data_agg
# Ensure column names in expression data match metadata
all(colnames(exp) %in% meta$sampleid)
all(colnames(exp) == meta$sampleid)
# Reorder expression data to match metadata
exp <- exp[, meta$sampleid]
# Round expression values and factorize conditions
exp <- round(exp)
meta$Condition <- factor(meta$Condition, levels = c("Control", "FBZ"))
class(meta$Condition)  # Check the class of the Condition column
```

```{r Follow DEseq2}
# Create DESeqDataSet object
dds <- DESeqDataSetFromMatrix(
  countData = exp,
  colData = meta,
  design = ~ Condition)
# Filter out low-count genes
dds <- dds[rowSums(counts(dds)) >= 10,]

# Run DESeq2 analysis
dds_run <- DESeq(dds)
# Extract results for the contrast FBZ vs Control
res <- results(dds_run, contrast=c("Condition", "FBZ", "Control"), alpha=0.05)
res <- res[complete.cases(res),]  # Remove rows with NA values

summary(res)  # Summarize the results
plotMA(res)
```

[***When to Use Which normalization method (vst/normtransform)?***]{.underline}

**normTransform (log2):**

1.  Use for quick exploratory analysis or when computational efficiency is a priority.

2.  Use when you need interpretable log2-transformed values.

**VST (Variance Stabilizing Transformation):**

1.  Use for high-quality exploratory analysis, especially when variance stabilization is critical (e.g., PCA, heatmaps).

2.  Use when analyzing datasets with a wide range of expression levels.

## 8- Normalize data using VST

```{r Normalization}
row_count <- counts(dds, normalized=FALSE)
vsn_data <- vsn2(row_count)
vsn_norm <- exprs(vsn_data)
##Same as above 3 lines but in one command
vsn_norm2 <- counts(dds, normalized=FALSE) %>% vsn2 %>% exprs
#use 'meanSdPlot' to verify the fit.
library(hexbin)
meanSdPlot(vsn_norm)
#save the normalized data
write.csv(vsn_norm, "vsn_norm.csv")

# Alternatively, use normTransform for normalization
ntd <- normTransform(dds)
exp.norm <- assay(ntd)
meanSdPlot(exp.norm)
```

## 9- Extract DEGs (from DEseq2 object) and their normalized expression levels (from the normalized matrix we made)

```{r DEGs}
degs <- res[res$padj < 0.05 & res$log2FoldChange > 1,]
degs.genes <- rownames(degs)
degs.exp <- vsn_norm[degs.genes,]
write.csv(degs.exp, "degs.exp.csv")
```

## 10- Downstream plots

### 10- (a) Heatmap for top 100 DEGs

```{r Heatmap}
deg_100 <- degs[order(degs$padj, -abs(degs$log2FoldChange)),]   #sort to get top 100
deg_100 <- degs[1:100,]                                         #select only them
deg_100_exp <- vsn_norm[rownames(deg_100),]                     #get their normalized expression values

# Create column annotations for the heatmap
sam_condition <- meta$Condition
column_annot <- HeatmapAnnotation(
  Condition = sam_condition,
  col = list(Condition = c("Control" = "limegreen", "FBZ" = "orange")))

# Generate the heatmap
Heatmap(
  matrix = deg_100_exp,
  top_annotation = column_annot,
  row_title = "Top 100 DEGs",
  column_title = "Samples",
  cluster_rows = TRUE,
  cluster_columns = TRUE,
  show_row_names = TRUE,
  show_column_names = TRUE,
  row_names_gp = gpar(fontsize = 4),
  column_names_rot = 45,
  column_names_centered = TRUE
)
```

## 10- (b) Perform PCA on DEGs

```{r PCA again}
degs_normalized <- vsn_norm[rownames(vsn_norm) %in% rownames(degs),]
pca_result <- prcomp(t(degs_normalized), scale. = TRUE)
pca_data <-pca_result$x
pca_result$x
# Plot 2D PCA
autoplot(pca_result, data = meta, colour = 'Condition',frame = T,label = T, label.size = 3,shape="Condition")

# Plot 3D PCA
mycolors <- ifelse(meta$Condition == "FBZ", "red", "lightgreen")
plot3d(pca_result$x[, 1:3], col = mycolors, size = 12, type = "s", main = "3D PCA Plot")
```

## 10- (c) Volcano plot

```{r Volcano plot}
res_df <- data.frame(res)
res_df$gene_symbol <- rownames(res_df)

# Biostatsquid theme
theme_set(theme_classic(base_size = 20) +
            theme(
              axis.title.y = element_text(face = "bold", margin = margin(0,20,0,0), size = rel(1.1), color = 'black'),
              axis.title.x = element_text(hjust = 0.5, face = "bold", margin = margin(20,0,0,0), size = rel(1.1), color = 'black'),
              plot.title = element_text(hjust = 0.5)))
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2fc respectively positive or negative)
res_df$diffexpressed <- "NA"
# if log2Foldchange > 1 and pvalue < 0.05, set as "UP"
res_df$diffexpressed[res_df$log2FoldChange > 1 & res_df$pvalue < 0.05] <- "UP"
# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
res_df$diffexpressed[res_df$log2FoldChange < -1 & res_df$pvalue < 0.05] <- "DOWN"
# Explore a bit
head(res_df[order(res_df$padj) & res_df$diffexpressed == 'DOWN', ])
# Create a new column "delabel" to de, that will contain the name of the top 30 differentially expressed genes (NA in case they are not)
res_df$delabel <- ifelse(res_df$gene_symbol %in% head(res_df[order(res_df$padj), "gene_symbol"], 30), res_df$gene_symbol, NA)


ggplot(res_df, aes(x = log2FoldChange, y = -log(padj), col = res_df$diffexpressed),label=delabel) +
  geom_point(alpha = 0.4) +
  scale_color_manual(values = c("Not Significant" = "grey", "Significant" = "red")) +
  labs(
    title = "Volcano Plot",
    x = "log Fold change",
    y = "-log10 adjusted Pvalue"
  ) +
  geom_vline(xintercept = c(1, -1), col="black",linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col="black",linetype = "dashed")+
    geom_point(size = 2)+
    scale_color_manual(values = c("#00AFBB", "grey", "#FFDB6D"), # to set the colours of our variable
                     labels = c("Downregulated", "Not significant", "Upregulated"))+ # to set the labels in case we want to overwrite the categories from the dataframe (UP, DOWN, NO)
coord_cartesian(ylim = c(0, 150), xlim = c(-10, 10)) + # since some genes can have minuslog10padj of inf, we set these limits
  labs(color = 'Regulation', #legend_title, 
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) + 
  scale_x_continuous(breaks = seq(-10, 10, 2))+ # to customise the breaks in the x axis
ggtitle('Ovarian cells volcano plot cancer vs healthy patients')+  # Plot title 
geom_text_repel(label=res_df$delabel,max.overlaps = Inf) # To show all labels 

```

## 11- Save normalized expression data as text (used for gene enrichment analysis e.g. Enrichr)

```{r Save for PEA}
exp_norm <- read.csv("vsn_norm.csv", header = TRUE, row.names = 1)
write.table(exp_norm, file = "exp_norm.txt", sep = "\t", quote = FALSE, row.names = TRUE)
```
